CMakeLists.txt

```diff
@@ -737,11 +737,10 @@ ELSE ()
 ENDIF ()
 IF (BROTLI_FOUND AND NOT WITHOUT_LIBS)
     INSTALL(TARGETS libh2o-evloop DESTINATION ${CMAKE_INSTALL_LIBDIR})
-    INSTALL(TARGETS h2o-httpclient DESTINATION ${CMAKE_INSTALL_BINDIR})
 ELSE ()
     SET_TARGET_PROPERTIES(libh2o-evloop PROPERTIES EXCLUDE_FROM_ALL 1)
-    SET_TARGET_PROPERTIES(h2o-httpclient PROPERTIES EXCLUDE_FROM_ALL 1)
 ENDIF()
+INSTALL(TARGETS h2o-httpclient DESTINATION ${CMAKE_INSTALL_BINDIR})
```

Comment:
Considering that h2o-httpclient is merely a wrapper of libh2o-evloop with only one source file (src/httpclient.c), I think that they should either be installed together or not.

Also, by reading #2326, it seems to me that our intent has been to let libh2o / libh2o-evloop depend solely on libbrotli provided by the environment (rather than using the one under deps/brotli), and to install libh2o-evloop and h2o-httpclient only if libbrotli was available.

@gfx Am I correct to understand that we are observing regression from #2326?

Reply:
Let me clarify the current dependency state about brotli (assuming `libbrotli` is a brotli package installed out of the project):

```mermaid
graph TD
    A[h2o] -->|depends on| B[deps/brotli]
    D[libh2o] -->|depends on| E[libbrotli]
    F[h2o-httpclient] -->|depends on| D
```

Is that correct?

I'm not sure why this project bundles `deps/brotli` while it depends on `libbrotli` for some artifacts at the same time. It's not really a problem right now, but this concept will lead to another complex situation when h2o supports a new compression algorithm. 

So once away from this pull-request, I think it's ideal to remove `deps/brotli` from the h2o source tree and to use only `libbrotli` to support the brotli algorithm in order to simplify the dependencies. If one makes sure to use an up-to-date version of brotli, they can build and install it by themselves. In that situation, this pull-request will never be effective. 


What do you think?

Reply:
Thank you for the chart and the detailed explanation. They help us make progress.

I think the point what you are missing is that brotli is a mandatory dependency of h2o as a standalone server. As discussed in #2326, you cannot disable certificate compression in h2o. At the same time, brotli has to be an external dependency of libh2o, as users of libraries do not expect functions of one library to be part of another library.

Based on these points, I think we have three options:
* a) continue with the current approach,
* b) remove `deps/brotli` and make libbrotli a mandatory dependency of h2o and libh2o,
* c) remove `deps/brotli` and make libbrotli a conditional dependency of h2o.

In order to adopt option b, I think we have to check that libbrotli is easily available on all the platforms that we currently care of (linux, macos, freebsd, netbsd). The downside of option c is that the change might be disruptive for existing users of h2o; when they try to upgrade, certificate compression might stop working.

Are you interested in running the neressary research for option b? (generally speaking I think it is a good idea to have access to these environment either as physical machines or as VMs).

Reply:
> I think the point what you are missing is that brotli is a mandatory dependency of h2o as a standalone server. As discussed in https://github.com/h2o/h2o/pull/2326, you cannot disable certificate compression in h2o. At the same time, brotli has to be an external dependency of libh2o, as users of libraries do not expect functions of one library to be part of another library.

Right, I missed the fact that brotli is required to support certificate compression. So it totally makes sense to have two different ways to use brotli. I think the option b and c are virtually impossible, as we don't know which OS is actually used to run h2o. 

So I propose another option:

d) In addition to the current approach, add an option to use external libbrotli to build the standalone h2o, omitting to build `deps/brotli`.

The downside of the option d is that it makes `CMakeLists.txt` more complicated. It might be worth considering because it could reduce the build time of h2o. However, I'm inclined not to do it, as a complicated `CMakeLists.txt` would lead to another problem.

Thank you for the discussion. It has clarified the attitude to compression libraries. It also makes it clear when h2o supports another compression library -- unless it's required to support a mandatory feature like certificate compression, it should be an optional dependency.